File: two-sum/river20s.java
/*
 * T.C: O(n) -> 배열 nums를 한 번 순회
 * S.C: O(n) → 최대 n개의 요소가 저장됨
 */
import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {

        HashMap<Integer, Integer> map = new HashMap<>();
        for  (int i = 0; i < nums.length; i++) {
            int k = target - nums[i];

            if (map.containsKey(k)) {
                return new int[] { map.get(k), i };
            }

            map.put(nums[i], i);
        }
        throw new IllegalArgumentException("exception handling for java compilation");
        
    }
}

File: two-sum/river20s.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        seen = {}
        for i, num in enumerate(nums):
            comp = target - num
            if comp in seen:
                return seen[comp], i
            if num not in seen:
                seen[num] = i
        return None

